File: apps/bank/authorization/go/authorization.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net/url"
	"os"
	"strings"

	"github.com/golang-jwt/jwt/v5"
	"github.com/valyala/fasthttp"
)

var (
	headersRemove = map[string]struct{}{
		"ce-id":          {},
		"ce-specversion": {},
		"ce-type":        {},
		"ce-source":      {},
		"content-type":   {},
		"host":           {},
		"x-k-sink":       {},
	}
	secretKey = os.Getenv("SECRET_KEY")
	kSink     = os.Getenv("K_SINK")
)

func main() {
	if secretKey == "" {
		log.Fatal("Missing required environment variable: SECRET_KEY")
	}

	server := fasthttp.Server{
		Handler: requestHandler,
	}

	log.Println("Server listening on port 8080...")
	if err := server.ListenAndServe(":8080"); err != nil {
		log.Fatalf("Error starting server: %v", err)
	}
}

func requestHandler(ctx *fasthttp.RequestCtx) {
	authHeader := string(ctx.Request.Header.Peek("Authorization"))
	if !strings.HasPrefix(strings.ToLower(authHeader), "bearer ") {
		forwardToBroker(ctx, nil, "Missing or invalid token", 401)
		return
	}
	tokenStr := authHeader[7:]
	
	claims := jwt.MapClaims{}
	token, err := jwt.ParseWithClaims(tokenStr, claims, func(token *jwt.Token) (interface{}, error) {
		return []byte(secretKey), nil
	})

	if err != nil || !token.Valid {
		forwardToBroker(ctx, nil, "Invalid token", 401)
		return
	}

	client, _ := claims["sub"].(string)
	hasAccess, _ := claims["has_access"].(bool)
	if !hasAccess {
		forwardToBroker(ctx, &client, "Access denied", 403)
	} else {
		forwardToBroker(ctx, &client, "Forwarded", 200)
	}
}

func forwardToBroker(ctx *fasthttp.RequestCtx, client *string, message string, statusCode int) {
	proceed := (statusCode == 200)
	
	var payload map[string]interface{}
	if err := json.Unmarshal(ctx.Request.Body(), &payload); err != nil {
		payload = make(map[string]interface{})
	}

	payload["client"] = client
	payload["message"] = message
	
	outHeaders := make(map[string]string)
	ctx.Request.Header.VisitAll(func(key, value []byte) {
		k := strings.ToLower(string(key))
		if _, exists := headersRemove[k]; !exists {
			outHeaders[string(key)] = string(value)
		}
	})

	if ceID := ctx.Request.Header.Peek("Ce-Id"); len(ceID) > 0 {
		outHeaders["Ce-Id"] = string(ceID)
	}
	outHeaders["Ce-Specversion"] = "1.0"
	outHeaders["Ce-Type"] = "authorization"
	outHeaders["Ce-Source"] = "authorization"
	outHeaders["Content-Type"] = "application/json"
	outHeaders["Ce-dv"] = fmt.Sprintf("%t", proceed)

	targetSink := string(ctx.Request.Header.Peek("X-K-Sink"))
	if targetSink == "" {
		targetSink = kSink
	}

	if targetSink != "" {
		parsedURL, err := url.Parse(targetSink)
		if err == nil && parsedURL.Host != "" {
			forwardEvent(parsedURL, outHeaders, payload)
		} else {
			log.Println("Warning: No valid sink configured. Event not forwarded.")
		}
	}

	ctx.SetStatusCode(statusCode)
	ctx.Response.Header.Set("Content-Type", "application/json")
	responseBody, _ := json.Marshal(map[string]string{"message": message})
	ctx.Write(responseBody)
}

func forwardEvent(parsedURL *url.URL, headers map[string]string, payload map[string]interface{}) {
	path := parsedURL.Path
	if path == "" {
		path = "/"
	}

	body, _ := json.Marshal(payload)
	
	r := fasthttp.AcquireRequest()
	r.SetRequestURI(parsedURL.String())
	r.Header.SetMethod("POST")
	for k, v := range headers {
		r.Header.Set(k, v)
	}
	r.SetBody(body)

	resp := fasthttp.AcquireResponse()
	err := fasthttp.Do(r, resp)
	if err != nil {
		log.Printf("Failed to forward event to %s: %v", parsedURL.String(), err)
	} else {
		log.Printf("Successfully forwarded event to: %s. Response: %s", parsedURL.String(), resp.Body())
	}

	fasthttp.ReleaseRequest(r)
	fasthttp.ReleaseResponse(resp)
}

File: apps/bank/authorization/php/authorization.php
<?php
use Swoole\Http\Server;
use Swoole\Http\Request;
use Swoole\Http\Response;
use Firebase\JWT\JWT;
use Firebase\JWT\Key;

require 'vendor/autoload.php';

// Define constant for headers that should not be forwarded.
const HEADERS_REMOVE = ["Ce-Id", "Ce-Specversion", "Ce-Type", "Ce-Source", "Content-Type", "Host", "X-K-Sink"];

$secretKey = getenv('SECRET_KEY');
$kSink = getenv('K_SINK');

if (!$secretKey) {
    echo "Missing required environment variable: SECRET_KEY\n";
    exit(1);
}

$server = new Server("0.0.0.0", 8080);

$server->set([
    'http_compression' => false,
]);

$server->on("request", function (Request $request, Response $response) use ($secretKey, $kSink) {
    // Retrieve the Authorization header
    $authHeader = $request->header['authorization'] ?? '';
    if (!$authHeader || stripos($authHeader, 'Bearer ') !== 0) {
        forwardToBroker($request, $response, null, "Missing or invalid token", 401);
        return;
    }
    $token = substr($authHeader, 7);

    try {
        $decoded = JWT::decode($token, new Key($secretKey, 'HS256'));
        $decodedArray = (array)$decoded;
        $client = $decodedArray['sub'] ?? "unknown";
        $hasAccess = $decodedArray['has_access'] ?? false;
        if (!$hasAccess) {
            forwardToBroker($request, $response, $client, "Access denied", 403);
        } else {
            forwardToBroker($request, $response, $client, "Forwarded", 200);
        }
    } catch (Exception $e) {
        forwardToBroker($request, $response, null, "Invalid token", 401);
    }
});

function forwardToBroker(Request $request, Response $response, ?string $client, string $message, int $statusCode)
{
    global $kSink;
    
    // Merge the original request JSON (if any) with additional fields.
    $bodyContent = $request->getContent();
    $originalBody = json_decode($bodyContent, true);
    $payload = is_array($originalBody) ? $originalBody : [];
    $payload['client'] = $client;
    $payload['message'] = $message;
    
    $proceed = ($statusCode === 200);

    // Build fixed outgoing headers.
    $outHeaders = [];

    // Copy additional headers from the incoming request, skipping those in HEADERS_REMOVE.
    foreach ($request->header as $key => $value) {
        $skip = false;
        foreach (HEADERS_REMOVE as $remove) {
            if (strcasecmp($key, $remove) === 0) {
                $skip = true;
                break;
            }
        }
        if (!$skip) {
            $outHeaders[$key] = $value;
        }
    }

    if (!empty($request->header['ce-id'])) {
        $outHeaders['Ce-Id'] = $request->header['ce-id'];
    }
    $outHeaders['Ce-Specversion'] = '1.0';
    $outHeaders['Ce-Type'] = 'authorization';
    $outHeaders['Ce-Source'] = 'authorization';
    $outHeaders['Content-Type'] = 'application/json';
    $outHeaders['Ce-dv'] = $proceed ? 'true' : 'false';

    // Determine target sink: use the "x-k-sink" header if present; otherwise, use the global K_SINK.
    $targetSink = (isset($request->header['x-k-sink']) && !empty($request->header['x-k-sink'])) 
                    ? $request->header['x-k-sink'] 
                    : $kSink;

    // Forward the payload if a valid target sink is provided.
    if ($targetSink) {
        $parsedUrl = parse_url($targetSink);
        $host = $parsedUrl['host'] ?? '';
        $path = $parsedUrl['path'] ?? '/';
        $port = $parsedUrl['port'] ?? 80;
        $client = new Swoole\Coroutine\Http\Client($host, $port);
        $client->setHeaders($outHeaders);
        $client->post($path, json_encode($payload));
        echo "Successfully forwarded event to: $targetSink. Response: " . $client->body . "\n";
    } else {
        echo "Warning: No valid sink configured. Event not forwarded.\n";
    }

    // Respond to the original requester.
    $response->status($statusCode);
    $response->header("Content-Type", "application/json");
    $response->end(json_encode(["message" => $message]));
}

$server->start();
